src/box/applier.cc

```diff
@@ -296,24 +298,30 @@ process_nop(struct request *request)
 }
 
 static int
-apply_row(struct xrow_header *row)
+apply_request(struct request *request)
 {
-	struct request request;
-	assert(!iproto_type_is_synchro_request(row->type));
-	if (xrow_decode_dml(row, &request, dml_request_key_map(row->type)) != 0)
-		return -1;
-	if (request.type == IPROTO_NOP)
-		return process_nop(&request);
-	struct space *space = space_cache_find(request.space_id);
+	if (request->type == IPROTO_NOP)
+		return process_nop(request);
+	struct space *space = space_cache_find(request->space_id);
 	if (space == NULL)
 		return -1;
-	if (box_process_rw(&request, space, NULL) != 0) {
-		say_error("error applying row: %s", request_str(&request));
+	if (box_process_rw(request, space, NULL) != 0) {
+		say_error("error applying row: %s", request_str(request));
 		return -1;
 	}
 	return 0;
 }
 
+static int
+apply_row(struct xrow_header *row)
```

Comment:
I don't like that now we have two sets of `apply` methods, one for decoded rows (`apply_request`) another for raw rows (`apply_row`). The `decoded` bool argument used to differentiate between the two cases makes the code difficult to follow IMO. AFAIU `apply_row` is needed solely for the FINAL JOIN phase. I'd recommend decoding rows somewhere in `applier_read_tx` or, even better, reusing the code applying rows between SUBSCRIBE and FINAL JOIN phases.

Reply:
Now applier_read_tx always decodes the incoming rows.